stempy/image.cpp

```diff
@@ -269,6 +270,62 @@ vector<STEMImage> createSTEMImages(InputIt first, InputIt last,
   return images;
 }
 
+// returns the binding result
+std::vector<double> getContainer(const STEMImage& inImage, const int numBins)
+{
+  // information about input STEMImage
+  int width = inImage.width;
+  int height = inImage.height;
+  auto curData = inImage.data;
+  std::cout << "Input STEM Image has width = " << width
+            << ", height = " << height << std::endl;
+
+  auto result = std::minmax_element(curData.get(), curData.get() + width * height);
+  double min = *result.first;
+  double max = *result.second;
+
+  // the "length" of each slot of the container
+  double length = static_cast<double>((max - min) / numBins);
+
+  std::vector<double> container;
+  // push all the intermediate values
+  for (int i = 0; i <= numBins; i++) {
+    container.push_back(min + i * length);
+  }
+
+  return container;
+}
+
+// function that computes histogram for all the STEM images
+// each histogram is a vector<int>
+std::vector<int> createSTEMHistogram(const STEMImage& inImage, const int numBins, const std::vector<double> bins)
+{
+  // initialize output
+  std::vector<int> frequencies(numBins, 0);
+
+  // STEMImage info
+  int width = inImage.width;
+  int height = inImage.height;
+  auto curData = inImage.data;
+  
+  // get a histrogram
+  for (int i = 0; i < width * height; i++) {
+    auto value = curData[i];
+    // check which bin it belongs to
+    for (int i = 0; i < numBins; i++) {
+      if (value >= bins[i] && value < bins[i + 1]) {
+        frequencies[i] += 1;
+      }
+      // the max value is put in the last slot
+      else if (value == bins[numBins]) {
```

Comment:
This shouldn't be necessary with the above suggested change to the binning.

Reply:
My question becomes - are you sure this is necessary? I guess you are trying to cover the corner case as it is `<` rather than `<=`. It is a little odd to see the spike at the end of the histograms you showed too, is that real, I wouldn't expect it to be.

Reply:
I think it is necessary. Although it would not give you bugs if I just change the interval to be both-side included, I feel like it produces uncertainty to values happen to be equal to the boundary value. Because technically it can belong to either interval. A clear classification makes me feel a bit more comfortable. But I agree that it is a little odd in the end to separately include the max-corner-case.

Reply:
I have fixed the bug, the max case should be outside the for loop which determines which bin the current value it belongs to. The weird artifact was due to the max being added 100 times.

Reply:
Thank you very much for catching the bug! @cryos

Reply:
As a side note on this: notice as well that a direct comparison is being made between a `uint64_t` (`value`) and a `double` (`bins[numBins]` - whose value is a result of arithmetic). I am not sure if this would be an issue or not, but thought I would point it out.